airflow/models/taskinstance.py

```diff
@@ -528,25 +532,43 @@ def _get_previous_ti(self, session=None):
                 if not previous_scheduled_date:
                     return None
 
-                return TaskInstance(task=self.task,
-                                    execution_date=previous_scheduled_date)
+                return TaskInstance(task=self.task, execution_date=previous_scheduled_date)
 
             dr.dag = dag
-            if dag.catchup:
-                last_dagrun = dr.get_previous_scheduled_dagrun(session=session)
+            if dag.catchup is False or state is not None or dag.schedule_interval is None:
```

Comment:
Can you also confirm that the tests cover all three cases? (It's early and my ☕️  hasn't kicked in yet).

Reply:
Yeah I will add comment.  I don't like it but it's only because for whatever reason current state is to use `get_previous_scheduled_dagrun` when `dag.catchup=True`.  `get_previous_scheduled_dagrun` is a little more efficient because it just gets one row instead of order by limit 1.   But when we are filtering by state we don't care about schedule and same when there _is_ no schedule.

Reply:
Re coverage yes these cases are covered that's what this part is doing:
```        schedule_interval_values = ('0 0 * * *', None, datetime.timedelta(days=1))
        catchup_values = (True, False)

        for tuple_ in product(schedule_interval_values, catchup_values):
            self._test_previous_dates(*tuple_, object_='previous_ti')
```
It's taking all combinations of the two lists and running the relevant tests for each combination.

Reply:
> The reasoning behind this `if` is hard to work out - could you expand on it with a comment saying why?

This is what I have so far:
```
# When `state` is given, we use `get_previous_dagrun` because we don't care about previous schedule: we
# just want prior dagrun with given `state`.
# When `schedule_interval is None`, we can't use schedule to find prior dag run.
# When `catchup=True`, we use `get_previous_scheduled_dagrun` for legacy reasons.
```

With status quo, if `catchup is True`, `previous_ti` is "last ti according to schedule, if one exists", but if `catchup is False`, previous_ti is "last ti absolutely, if one exists".  

When we do the spring cleaning of context params in 2.0, we should fix this and call things what they are, and not have behavior subtly change based on other params if it can be avoided.  

E.g. I would say `previous_ti` should _always_ give the absolute last ti without regard to schedule, and we should make a separate property `previous_scheduled_ti` for the last _scheduled_ ti.  Then these methods would behave the same no matter `catchup` or `schedule_interval`; one would call prev dag run, the other would call prev scheduled dag run.  I would be happy to separate these in this manner in a subsequent PR if you like this idea.

##### Note on the `schedule is None` and `catchup=True` case.  
In status quo, `previous_ti` _always_ returns `None`, because it is using `get_previous_scheduled_dagrun` when there is no schedule to use.  I discovered this when adding tests for `previous_ti`.  This PR fixes that by using `get_previous_dagrun` in this case.

Please LMK thoughts!   In particular:
* Are these comments sufficient?
* Do you like the idea of splitting previous_ti (absolute) and previous_scheduled_ti (scheduled)?

Reply:
OK this is actually a bit more readable:
```
# We always ignore schedule in dagrun lookup when `state` is given or `schedule_interval is None`.
# For legacy reasons, when `catchup=True`, we use `get_previous_scheduled_dagrun` unless `ignore_schedule`.
ignore_schedule = state is not None or dag.schedule_interval is None
if dag.catchup is True and not ignore_schedule:
    last_dagrun = dr.get_previous_scheduled_dagrun(session=session)
else:
    last_dagrun = dr.get_previous_dagrun(session=session, state=state)
```

Reply:
> E.g. I would say `previous_ti` should always give the absolute last ti without regard to schedule, and we should make a separate property `previous_scheduled_ti` for the last scheduled ti.

That sounds sensible for 2.0, yes. Can you reply to the post on the mailing list about context params with this change/suggestion? (I'm planning to cherry pick this PR to 1.10.4 once we get it merged, so we can't change the behaviour yet)

Comments LGTM